### PR TITLE
Low-light Image Enhancement using MIRNet

### PR DESCRIPTION
**Pull request by soumik12345**
With the goal of recovering high-quality image content from its degraded version, image restoration enjoys numerous applications, such as in surveillance, computational photography, medical imaging, and remote sensing. In this example, we implement the **MIRNet** model for low-light image enhancement, a fully convolutional architecture that learns an enriched set of features that combines contextual information from multiple scales, while simultaneously preserving the high-resolution spatial details.

Please find the corresponding notebook [here](https://colab.research.google.com/drive/1VtRwhq5sG_K3MNY6KYygtvUVC85Ju2qq?usp=sharing).

### CODE REVIEW

**Comment by soumik12345** [comment]
Hi @fchollet , can you please review this example?

File: examples/vision/mirnet.py [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+"""
+## Inference
+
+For inference, we need the dimensions of the input image to be multiples of 4. Hence we
+resize the input image to the closest possible dimension that is acceptable by the model
+before inference.
+"""
+
+
+def closest_number(n, m):
```
Comment (fchollet) [inline examples/vision/mirnet.py:457]: Please add a one-line description of this function

**Review by fchollet — COMMENTED** [review]
Looking great, thanks for the update!

File: examples/vision/mirnet.py [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+### Inference on Train Images
+"""
+
+for low_light_image in random.sample(train_low_light_images, 4):
+    original_image = Image.open(low_light_image)
+    enhanced_image = infer(original_image)
+    plot_result(original_image, enhanced_image)
+
+"""
+### Inference on Validation Images
```
[the remark this answers is not in the document]
Comment (soumik12345) [reply examples/vision/mirnet.py:480]: For all the test images on LOLDataset, I added the corresponding enhanced images produced by a free image brightness correction tool from [pinetools](https://pinetools.com/change-image-brightness) in order to compare with the results achieved by MIRNet. Comparisons on the test images can be found in the [Colab Notebook](https://colab.research.google.com/drive/1VtRwhq5sG_K3MNY6KYygtvUVC85Ju2qq?usp=sharing).

File: examples/vision/mirnet.py [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+
+"""shell
+!gdown https://drive.google.com/uc?id=157bjO1_cFuSd0HWDUuAmcHRJDVyWpOxB
+!unzip -q LOLdataset.zip
+"""
+
+"""
+## Creating a TensorFlow Dataset
+
+In order to restrict the training time to an hour, we would use 300 image pairs from the
```
[the remark this answers is not in the document]
Comment (soumik12345) [reply examples/vision/mirnet.py:54]: Used 300 images for training, remaining 185 for validation.

File: examples/vision/mirnet.py [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+
+"""
+## MIRNet Model
+
+Main features of the MIRNet model are:
+
+- A feature extraction model that obtains a complementary set of features across multiple
+spatial scales, while maintaining the original high-resolution features to preserve
+precise spatial details.
+
```
[the remark this answers is not in the document]
Comment (soumik12345) [reply examples/vision/mirnet.py]: Removed line breaks in between list items.

File: examples/vision/mirnet.py [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+def selective_kernel_feature_fusion(
+    multi_scale_feature_1, multi_scale_feature_2, multi_scale_feature_3
+):
+    channels = list(multi_scale_feature_1.shape)[-1]
+    combined_feature = layers.Add()(
+        [multi_scale_feature_1, multi_scale_feature_2, multi_scale_feature_3]
+    )
+    gap = layers.GlobalAveragePooling2D()(combined_feature)
+    channel_wise_statistics = tf.reshape(gap, shape=(-1, 1, 1, channels))
+    compact_feature_representation = layers.ReLU()(
```
[the remark this answers is not in the document]
Comment (soumik12345) [reply examples/vision/mirnet.py]: Passed `activation="relu"` to the Conv2D layer

File: examples/vision/mirnet.py [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+        layers.Conv2D(filters=channels // 8, kernel_size=(1, 1))(
+            channel_wise_statistics
+        )
+    )
+    feature_descriptor_1 = tf.nn.softmax(
```
[the remark this answers is not in the document]
Comment (soumik12345) [reply examples/vision/mirnet.py]: Passed `activation="softmax"` to the Conv2D layer

File: examples/vision/mirnet.py [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+
+
+def channel_attention_block(input_tensor):
+    channels = list(input_tensor.shape)[-1]
+    average_pooling = layers.GlobalAveragePooling2D()(input_tensor)
+    feature_descriptor = tf.reshape(average_pooling, shape=(-1, 1, 1, channels))
+    feature_activations = layers.ReLU()(
+        layers.Conv2D(filters=channels // 8, kernel_size=(1, 1))(feature_descriptor)
+    )
+    feature_activations = tf.nn.sigmoid(
```
[the remark this answers is not in the document]
Comment (soumik12345) [reply examples/vision/mirnet.py]: Passed `activation="sigmoid"` to the Conv2D layer

File: examples/vision/mirnet.py (hunk 1 above)
Comment (soumik12345) [reply examples/vision/mirnet.py]: Cleaned up the inference code and removed the unnecessary `closest_number` function.

**Comment by soumik12345** [comment]
Hi @fchollet , thanks for the suggestions, I have updated the example accordingly. Can you please review?

**Comment by soumik12345** [comment]
Updates:

- Added enhanced versions of the 15 test images from LOLDataset using an online tool called [pinetools](https://pinetools.com/change-image-brightness).
- Using 300 images for training, 185 for validation, and 15 as test images.
- Removed line breaks in between list items.
- Added `activation="relu"` in the Conv2D layers.
- Added `activation="softmax"` in the Conv2D layers.
- Added `activation="sigmoid"` in the Conv2D layers.
- Removed unnecessary functions in the inference.

File: examples/vision/mirnet.py [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+### Inference on Test Images
+
+We compare the test images from LOLDataset enhanced by MIRNet against the image
+brightness correction feature available for
+[Pinetools](https://pinetools.com/change-image-brightness).
+"""
+
+# Test Images in LOLDataset enhanced using automated brightness correction
+# option in Pinetools (https://pinetools.com/change-image-brightness)
+test_pinetools_enhanced_images = sorted(glob("./lol_dataset/eval15/pinetools/*"))
```
Comment (fchollet) [inline examples/vision/mirnet.py:498]: This is not a fair comparison, the enhancement here is extremely subpar. You can implement contrast normalization in pure Python or TensorFlow and it will work much better. Or just use PIL's `autocontrast`. The results will be much better and the comparison with the deep learning model will be fair.

File: examples/vision/mirnet.py [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+    output_image = output_image.reshape(
+        (np.shape(output_image)[0], np.shape(output_image)[1], 3)
+    )
+    output_image = Image.fromarray(np.uint8(output_image))
+    original_image = Image.fromarray(np.uint8(original_image))
+    return output_image
+
+
+"""
+### Inference on Train Images
```
Comment (fchollet) [inline examples/vision/mirnet.py:471]: You can remove this section, inference is only valuable on test or validation sets

File: examples/vision/mirnet.py [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+"""
+
+for low_light_image in random.sample(train_low_light_images, 4):
+    original_image = Image.open(low_light_image)
+    enhanced_image = infer(original_image)
+    plot_results([original_image, enhanced_image], ["Original", "Enhanced"])
+
+"""
+### Inference on Validation Images
```
Comment (fchollet) [inline examples/vision/mirnet.py:480]: You can also remove this section since you have a test section below

File: examples/vision/mirnet.py (hunk 8 above)
Comment (soumik12345) [reply examples/vision/mirnet.py]: Added comparison of the test images enhanced by MIRNet against the corresponding images enhanced using `PIL.ImageOps.autocontrast`.

File: examples/vision/mirnet.py (hunk 9 above)
Comment (soumik12345) [reply examples/vision/mirnet.py]: Removed inference on train images.

File: examples/vision/mirnet.py (hunk 10 above)
Comment (soumik12345) [reply examples/vision/mirnet.py]: Removed inference on validation images.

**Comment by soumik12345** [comment]
Hi @fchollet , thanks for the suggestions. I have updated the example accordingly. Also updated the corresponding [Colab Notebook](https://colab.research.google.com/drive/1VtRwhq5sG_K3MNY6KYygtvUVC85Ju2qq?usp=sharing). Can you please review it?

- Added comparison of the test images enhanced by MIRNet against the corresponding images enhanced using `PIL.ImageOps.autocontrast`.
- Removed inference on train images.
- Removed inference on validation images.

**Review by fchollet — APPROVED** [review]
Excellent, it looks great! The comparison with `autocontrast()` does a great job at demonstrating the value of the MIRNet approach.

I pushed some copyedits, please pull them first. After pulling the copyedits, please add the generated files. I believe we're ready to publish 👍

**Comment by soumik12345** [comment]
Hi @fchollet , added the generated files.

**Review by fchollet — APPROVED** [review]

**Comment by fchollet** [comment]
Thanks for the great contribution! 👍
